### ISSUE TITLE
False warning of unused variable

### ISSUE THREAD

**Issue by JerryKwan** [rust-lang/rust-clippy#11711]
### Summary

There is a false warning of unused variable in the following codes
```rust
let index_id = index_ids[i].clone();
let index_uid = index_uids[i].clone();
assert!(
    matches!(error, MetastoreError::NotFound(EntityKind::Index { index_id: index_id }) if index_id == index_uid.index_id())
);
```
Apparently the variable of index_id is used in the following sentence
But why it still complains unused?

### Reproducer

I tried this code:

```rust
let index_id = index_ids[i].clone();
let index_uid = index_uids[i].clone();
assert!(
    matches!(error, MetastoreError::NotFound(EntityKind::Index { index_id: index_id }) if index_id == index_uid.index_id())
);
```

I saw this happen:

```
|
669 |             let index_id = index_ids[i].clone();
    |                 ^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_index_id`
    |
    = note: requested on the command line with `-D unused-variables`

```

I expected to see this happen:
No unused variable warning

**Comment by JerryKwan** [comment]
Seems like the problem occurs in processing micro matches!()

**Comment by y21** [comment]
The lint looks like it's right. That `EntityKind::Index { index_id: index_id }` creates a new variable for use within the `matches!` and only that one is then used.
This doesn't use the other `index_id` variable, so it (correctly) flags it as unused.

Also, this is a rustc lint (clippy just extends the compiler with more lints), so this should be filed on the rust-lang/rust repo

**Comment by JerryKwan** [comment]
@y21 Thanks for looking into this issue
EntityKind::Index { index_id: index_id } use the outer index_id to build EntiryKind::Index, right?

**Comment by Patryk27** [comment]
>  EntityKind::Index { index_id: index_id } use the outer index_id to build EntiryKind::Index, right?

When building a struct, yes - but not when matching:
```rust
struct Foo {
    bar: i32,
}

fn main() {
    let bar = 123;
    
    // Uses variable `bar` to construct `Foo`:
    let foo = Foo { bar };
    
    let bar = 321;

    // Creates a new variable `bar` containing value from `foo.bar` - i.e. this
    // matches *always*:
    if let Foo { bar } = foo {
        println!("{}", bar);
    }
}
```

**Comment by JerryKwan** [comment]
@Patryk27 
Thanks for the detailed explanation, thank you
